### PR TITLE
[2201.3.x] Fix regular expressions not working with table keys and query expressions/actions

### PR DESCRIPTION
**Pull request by lasinicl**
## Purpose
$title

Fixes https://github.com/ballerina-platform/ballerina-lang/issues/38692
Fixes https://github.com/ballerina-platform/ballerina-lang/issues/38679

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Comment by KavinduZoysa** [comment]
@lasinicl, shall we add few test cases, that create tables and maps using query expressions?

`query-expr := [query-construct-type] query-pipeline select-clause [on-conflict-clause]`

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/DataflowAnalyzer.java [hunk 1]
```diff
@@ -1290,12 +1301,90 @@ public Integer hash(Node node) {
...
+        return regexpHash;
+    }
+
+    private Integer generateHashForReCharacterClass(BLangReCharacterClass reCharacterClass) {
+        int regexpHash = hash(reCharacterClass.characterClassStart) + hash(reCharacterClass.characterClassEnd);
+        if (reCharacterClass.negation != null) {
+            regexpHash = regexpHash + hash(reCharacterClass.negation);
+        }
+        if (reCharacterClass.charSet == null) {
+            return regexpHash;
```
Comment (prakanth97) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/DataflowAnalyzer.java:1359]: Shall we add a test to cover this?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/DataflowAnalyzer.java [hunk 2]
```diff
@@ -1290,12 +1301,90 @@ public Integer hash(Node node) {
...
+            }
+        }
+        return regexpHash;
+    }
+
+    private Integer generateHashForReCapturingGroup(BLangReCapturingGroups reCapturingGroup) {
+        int regexpHash = hash(reCapturingGroup.openParen) +
+                generateHashForRegExp(reCapturingGroup.disjunction.sequenceList) + hash(reCapturingGroup.closeParen);
+        if (reCapturingGroup.flagExpr == null) {
+            return regexpHash;
```
Comment (prakanth97) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/DataflowAnalyzer.java:1378]: Shall we add a test with recapturingGroup without flags as table key

File: tests/jballerina-unit-test/src/test/resources/test-src/types/table/table_key_field_value_test.bal [hunk 3]
```diff
@@ -467,6 +467,44 @@ function testKeyCollision() {
...
+    if (err2 is error) {
+        map<string> msg = {"message":"a value found for key 'AB*[^abc-efg](?:A|B|[ab-fgh]+(?im-x:[cdeg-k]??)|)|^|PQ?'"};
+        assertEqual(msg, err2.detail());
+    }
+}
+
+function getTable() returns table<Row9> key(k) {
+    return table [{k: re `AB*[^abc-efg](?:A|B|[ab-fgh]+(?im-x:[cdeg-k]??)|)|^|PQ?`, value: 17},
+                  {k: re `AB*[^abc-efg](?:A|B|[ab-fgh]+(?im-x:[cdeg-k]??)|)|^|PQ?`, value: 17}];
+}
```
Comment (prakanth97) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/table/table_key_field_value_test.bal:506]: Shall we add few test for table with interpolation in regular expression?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/DataflowAnalyzer.java [hunk 4]
```diff
@@ -1290,12 +1301,90 @@ public Integer hash(Node node) {
...
+        } else if (reAtom.getKind() == NodeKind.REG_EXP_CAPTURING_GROUP) {
+            regexpHash = regexpHash + generateHashForReCapturingGroup((BLangReCapturingGroups) reAtom);
+        }
+        if (reAtomQuantifier.quantifier != null) {
+            BLangReQuantifier reQuantifier = reAtomQuantifier.quantifier;
+            regexpHash = regexpHash + hash(reQuantifier.quantifier);
+            if (reQuantifier.nonGreedyChar == null) {
+                return regexpHash;
+            }
+            return regexpHash + hash(reQuantifier.nonGreedyChar);
```
Comment (pcnfernando) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/DataflowAnalyzer.java:1348]:
```suggestion
            if (reQuantifier.nonGreedyChar != null) {
                return regexpHash + hash(reQuantifier.nonGreedyChar);
            }
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/DataflowAnalyzer.java [hunk 5]
```diff
@@ -1290,12 +1301,90 @@ public Integer hash(Node node) {
...
+    private Integer generateHashForReCapturingGroup(BLangReCapturingGroups reCapturingGroup) {
+        int regexpHash = hash(reCapturingGroup.openParen) +
+                generateHashForRegExp(reCapturingGroup.disjunction.sequenceList) + hash(reCapturingGroup.closeParen);
+        if (reCapturingGroup.flagExpr == null) {
+            return regexpHash;
+        }
+        BLangReFlagExpression flagExpr = (BLangReFlagExpression) reCapturingGroup.flagExpr;
+        regexpHash = regexpHash + hash(flagExpr.questionMark) + hash(flagExpr.colon);
+        if (flagExpr.flagsOnOff == null) {
+            return regexpHash;
```
Comment (pcnfernando) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/DataflowAnalyzer.java:1383]: Shall we add a test to cover this?

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/query/SimpleQueryExpressionWithDefinedTypeTest.java [hunk 6]
```diff
@@ -367,6 +368,21 @@ public void testUsingAnIntersectionTypeInQueryExpr() {
         BRunUtil.invoke(result, "testUsingAnIntersectionTypeInQueryExpr");
     }
 
+    @Test(dataProvider = "dataToTestQueryExpr")
+    public void testQueryExpr(String functionName) {
```
Comment (pcnfernando) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/query/SimpleQueryExpressionWithDefinedTypeTest.java:372]:
```suggestion
    public void testQueryExprWithRegExp(String functionName) {
```

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/query/SimpleQueryExpressionWithDefinedTypeTest.java [hunk 7]
```diff
@@ -367,6 +368,21 @@ public void testUsingAnIntersectionTypeInQueryExpr() {
...
+        BRunUtil.invoke(result, functionName);
+    }
+
+    @DataProvider
+    public Object[] dataToTestQueryExpr() {
```
Comment (pcnfernando) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/query/SimpleQueryExpressionWithDefinedTypeTest.java:377]:
```suggestion
    public Object[] dataToTestQueryExprWithRegExp() {
```

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/query/QueryExprWithQueryConstructTypeTest.java [hunk 8]
```diff
@@ -398,6 +398,22 @@ public void testMapConstructingQueryExprWithStringSubtypes() {
         BRunUtil.invoke(result, "testMapConstructingQueryExprWithStringSubtypes");
     }
 
+    @Test(dataProvider = "dataToTestQueryExprWithQueryConstructTyp")
```
Comment (pcnfernando) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/query/QueryExprWithQueryConstructTypeTest.java:401]:
```suggestion

    @Test(dataProvider = "dataToTestQueryExprWithQueryConstructType")
```

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/query/QueryExprWithQueryConstructTypeTest.java [hunk 9]
```diff
@@ -398,6 +398,22 @@ public void testMapConstructingQueryExprWithStringSubtypes() {
...
+    public void testQueryExprWithQueryConstructType(String functionName) {
+        BRunUtil.invoke(result, functionName);
+    }
+
+    @DataProvider
+    public Object[] dataToTestQueryExprWithQueryConstructTyp() {
```
Comment (pcnfernando) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/query/QueryExprWithQueryConstructTypeTest.java:407]:
```suggestion
    public Object[] dataToTestQueryExprWithQueryConstructType() {
```

File: tests/jballerina-unit-test/src/test/resources/test-src/query/query-expr-with-query-construct-type.bal [hunk 10]
```diff
@@ -1391,6 +1391,92 @@ function testMapConstructingQueryExprWithStringSubtypes() {
...
+    var t = arr2.iterator();
+    assertEqual({value: re `A`}, t.next());
+    assertEqual({value: re `C`}, t.next());
+    assertEqual((), t.next());
+}
+
+function testNestedQueryExprConstructingTableWithRegExp() {
+    string:RegExp[] arr1 = [re `A`, re `B2`, re `C`];
+    int v = 1;
+    table<Row> res = table key() from var re in (from string:RegExp reg in arr1
```
Comment (pcnfernando) [inline tests/jballerina-unit-test/src/test/resources/test-src/query/query-expr-with-query-construct-type.bal:1455]: Shall we recheck the formatting?

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TableUtils.java [hunk 11]
```diff
@@ -93,6 +94,8 @@ public static Long hash(Object obj, Node parent) {
                     result = 31 * result + hash(tableIterator.next(), node);
                 }
                 return result;
+            } else if (refValue instanceof RegExpValue) {
+                return (long) refValue.toString().hashCode();
```
Comment (gimantha) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TableUtils.java:98]: Can this line cause hash collision with a similar string value?

Comment (lasinicl) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TableUtils.java:98]: No. Added tests to cover this.

**Review by pcnfernando — APPROVED** [review]

**Pull request merged**